core/iwasm/libraries/lib-pthread/lib_pthread_wrapper.c

```diff
@@ -1085,6 +1122,169 @@ posix_memalign_wrapper(wasm_exec_env_t exec_env, void **memptr, int32 align,
     return 0;
 }
 
+#ifdef WASM_ENABLE_LIB_PTHREAD_SEMAPHORE
+
+static int32
+sem_open_wrapper(wasm_exec_env_t exec_env, const char *name, int32 oflags)
+{
+    korp_sem *psem = NULL;
+    ThreadInfoNode *info_node = NULL;
+
+    /**
+     * For RTOS, global semaphore map is safety for share the same semaphore
+     * between task/pthread.
+     * For Unix like system, it's dedicate for multiple processes.
+     */
+
+    if ((info_node = bh_hash_map_find(sem_info_map, (void *)name))) {
+        return info_node->handle;
+    }
+
+    if (!(psem = os_sem_open(name, oflags))) {
+        goto fail1;
+    }
+
+    if (!(info_node = wasm_runtime_malloc(sizeof(ThreadInfoNode))))
+        goto fail2;
+
+    memset(info_node, 0, sizeof(ThreadInfoNode));
+    info_node->exec_env = exec_env;
+    info_node->handle = allocate_handle();
+    info_node->type = T_SEM;
+    info_node->u.sem = psem;
+    info_node->status = SEM_CREATED;
+
+    if (!append_thread_info_node(info_node))
+        goto fail3;
+
+    if (!bh_hash_map_insert(sem_info_map, (void *)name, info_node)) {
+        goto fail3;
+    }
+
+    return info_node->handle;
+
+fail3:
+    delete_thread_info_node(info_node);
```

Comment:
Seems should be:
```C
fail4:
    delete_thread_info_node(info_node);
fail3:
    wasm_runtime_free(info_node);

Reply:
info_node was freed in delete_thread_info_node -> thread_info_destroy

Reply:
It seems work, but L1152 also goto fail3 and L1152 hasn't appended the node to thread_info_map yet, we can simply free the node. The code had better be like:
```C
    if (!(info_node = wasm_runtime_malloc(sizeof(ThreadInfoNode))))
        goto fail2;

    if (!append_thread_info_node(info_node)) {
        wasm_runtime_free(info_node);
        goto fail2;
    }

    if (!bh_hash_map_insert(sem_info_map, (void *)name, info_node))
        goto fail3;  //only goto fail3 when append thread info node success

    return info_node->handle;
```